tests/test_floss.py

```diff
@@ -411,3 +419,85 @@ def test_aamp_floss_inf_nan(substitute, substitution_locations):
             npt.assert_almost_equal(ref_P, comp_P)
             npt.assert_almost_equal(ref_I, comp_I)
             npt.assert_almost_equal(ref_T, comp_T)
+
+
+def test_floss_with_isconstant():
+    def custom_func(a, w):
+        _, Σ_a = core.compute_mean_std(a, w)
+
+        return Σ_a < 0.5
```

Comment:
The function `isconstant_func_stddev_threshold(T, m, ...)` decides a subsequence is constant or not based on its standard deviation and a threshold, denoted as `stddev_threshold` in the code. However, the value of this threshold depends on the full timeseries `T` (and not just one subsequence); See line below:

https://github.com/TDAmeritrade/stumpy/blob/203aa3f7721509d1139e451c43bb3aaccc63d713/tests/naive.py#L2088

A subequence `S` (in the timeseries `T`) has a certain value for its standard deviation; however, the threshold `stddev_threshold`, in the line shown above, DEPENDS on the stddev of ALL subseqs in `T`. So, when I update `T`, the threshold might change. In other words, `S` might be constant in `T`, but may not be constant in `np.append(T[1:], new_value)`. 


Now, let's see what is happening in `floss`:

```
# when `floss` object is initialized, we compute `self._T_subseq_isconstant`
self._T_subseq_isconstant = core.rolling_isconstant(self._T, self._m, self._T_subseq_isconstant_func)

# Then when we call `update()`, we have:
self._T_subseq_isconstant[:-1] = self._T_subseq_isconstant[1:]
self._T_subseq_isconstant[-1] = core.rolling_isconstant(self._T[-self._m :], self._m, self._T_subseq_isconstant_func)
```
Here, we shift the array so as to avoid computing the full array everytime. Note that the last line applies the function `self._T_subseq_isconstant_func` on ONLY ONE SUBSEQUENCE as opposed to the whole time series. Also note that the program DOES NOT MEMORIZE / LEARN the threshold `stddev_threshold` and instead it just compute it whenever the function `self._T_subseq_isconstant_func` is called. So, a different result can be obtained if, instead of shifting `self._T_subseq_isconstant`, we compute it for the whole data (and not just the last subsequence). In other words, the outcome of
```
self._T_subseq_isconstant = core.rolling_isconstant(self._T, self._m, self._T_subseq_isconstant_func)
```
can be different. Again, just to be clear, this is only because of the way the function `isconstant_func_stddev_threshold` is designed, where the stddev threshold depends on the sliding_stddev of the subsequences in the first argument of this function.

---

Now, if we look at the test function `test_floss.py`, we can see the following line:
```
ref_T_subseq_isconstant = custom_func(ref_T, m)
```
Here, we compute "subseq_isconstant " for the full array `ref_T`. (`ref_T` is a shifted version of data[:n], see the for loop ` for i, ref_T in enumerate(rolling_Ts):` in `the test_floss.py`) 

To wrap up, if I want to use the function `isconstant_func_stddev_threshold(T, m, ...)` then, I need to avoid shifiting `self._T_subseq_isconstant `, and instead just recompute it for the array as follows:
```
# in .update()
T_subseq_isconstant = core.rolling_isconstant(self._T, self._m, func)
```
note that `self._T` is the updated version of `T`. Also, I do not need to store `T_subseq_isconstant ` as attribute since we need to recompute the full array everytime. The downside is we may increase the computing time of `.update()`. Also, I am not sure even if it makes any sense to have a function to say a subsequence `S` is constant in `T` but then it says it is not constant in the shifted version of `T`. So, maybe it may not be meaningful to have a funciton like `isconstant_func_stddev_threshold`, where its output regarding a particular subsequence can be different between `T` and `np.append(T[1:], new_value)`.

Reply:
Thanks for the explanation. So, the issue seems to be with `isconstant_func_stddev_threshold` in that it requires knowledge of the FULL time series in order to compute the stddev threshold. Here are some thoughts/ideas:

Move `stddev_threshold` into the list of parameters of `isconstant_func_stddev_threshold` so that it looks like:
```
def isconstant_func_stddev_threshold(a, w, quantile_threshold=0, stddev_threshold=None):
    sliding_stddev = rolling_nanstd(a, w)
    if stddev_threshold is not None:
        return sliding_stddev <= stddev_threshold
    elif quantile_threshold == 0:  # pragma: no cover
        return sliding_stddev == 0
    else:
        stddev_threshold = np.quantile(sliding_stddev, quantile_threshold)
        return sliding_stddev <= stddev_threshold
```

Then, instead of computing `stddev_threshold` for the full time series, I would simply compute `stddev_threshold` ONCE AND ONLY ONCE for the initial `T` chunk and then, for each new data point and corresponding new subsequence, use the same pre-computed `stddev_threshold` from the first chunk. In other words, you make an assumption that all future, unseen data will have the same `stddev_threshold`.

To me, this is somewhat similar to ML feature engineering when you use [MinMaxScaler](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MinMaxScaler.html) to force your training data to have values between `0` and `1`. However, it is completely possible that your test set may have values that are larger than your `max` training data and therefore causes you to have test values that are larger than `1`. It's an assumption and I think it's okay especially for unit testing.

Reply:
> Then, instead of computing `stddev_threshold` for the full time series, I would simply compute `stddev_threshold` ONCE AND ONLY ONCE for the initial `T` chunk and then, for each new data point and corresponding new subsequence, use the same pre-computed `stddev_threshold` from the first chunk

Cool! I will revise it.

> To me, this is somewhat similar to ML feature engineering when you use [MinMaxScaler](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MinMaxScaler.html) to force your training data to have values between `0` and `1`. However, it is completely possible that your test set may have values that are larger than your `max` training data and therefore causes you to have test values that are larger than `1`. It's an assumption and I think it's okay especially for unit testing.

Exactly! We assume the initial `T` chunk represents the distribution of stddev for FULL data.